packages/backend/src/nest/storage/UserProfileStore.ts

```diff
@@ -0,0 +1,261 @@
+import { EventEmitter } from 'events'
+import KeyValueStore from 'orbit-db-kvstore'
+import OrbitDB from 'orbit-db'
+import { getCrypto, ICryptoEngine } from 'pkijs'
+import { sha256 } from 'multiformats/hashes/sha2'
+import { stringToArrayBuffer } from 'pvutils'
+import * as Block from 'multiformats/block'
+import * as dagCbor from '@ipld/dag-cbor'
+
+import { Logger } from '@quiet/logger'
+import { NoCryptoEngineError, UserProfile } from '@quiet/types'
+import { keyObjectFromString, verifyDataSignature } from '@quiet/identity'
+
+import { StorageEvents } from './storage.types'
+import createLogger from '../common/logger'
+
+const logger = createLogger('UserProfileStore')
+
+/**
+ * Check magic byte sequence to determine if buffer is a PNG image.
+ */
+export const isPng = (buffer: Uint8Array): boolean => {
+  // https://en.wikipedia.org/wiki/PNG
+  const pngHeader = [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]
+
+  if (buffer.length < pngHeader.length) {
+    return false
+  }
+
+  for (let i = 0; i < pngHeader.length; i++) {
+    if (buffer[i] !== pngHeader[i]) {
+      return false
+    }
+  }
+  return true
+}
+
+/**
+ * Takes a base64 data URI string that starts with 'data:*\/*;base64,'
+ * as returned from
+ * https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsDataURL
+ * and converts it to a Uint8Array.
+ */
+export const base64DataURLToByteArray = (contents: string): Uint8Array => {
+  const [header, base64Data] = contents.split(',')
```

Comment:
I think these are helpful to read:
- https://security.stackexchange.com/questions/135513/what-could-an-img-src-xss-do
- https://security.stackexchange.com/a/124843
- https://security.stackexchange.com/questions/258384/can-xss-ever-occur-in-an-img-tags-src-attribute

I don't think JS is allowed in a src tag in Chrome. With external HTTP requests, I think it's possible that someone does a CSRF, but since Quiet is running in it's own isolated browser, the user shouldn't have any session data for other sites. In this PR, we are checking that the binary data starts with the magic byte sequence of a PNG. Unless there is a way for the browser to interpret data starting with that magic byte sequence as some other type of data, I think it should be sufficient to start. Another thing might be adding a CSP and checking that the data url starts with `image/png`. If you can find any other resources that help us improve the safety of the application, we can apply the lessons we learn.

Reply:
I think the most important thing with the `img` tag and untrusted input is making sure the `src` URL is safe (e.g. actually points to an image). Otherwise, I trust the browser to safely display all images. Of course, it may not. But I don't think we need to be too concerned about that right now.

In order to check the `src` URL (which is user-supplied), we verify that it starts with `data:image/png;base64`. We additionally check the magic byte sequence of the payload which might be unnecessary.

I looked at Signal, but I didn't find much in relation to validating images. It does look like they use HTML canvas to resize images which might be a simpler alternative to ImageMagick.

Currently, the user supplies the `src` URL and we validate it. However, another option is to only send the image data instead of the full data URL and then we prepend the `data:image/png;base64,` part ourselves. Then we aren't validating that the URL is correct - we are creating the URL ourselves with the base64 encoded data that the user provides. So less chance that validation breaks.

Reply:
That sounds like a good idea! It gives the attacker less control, right?

Reply:
Not really, I think it's just about reducing risk of an if statement not working.

If you have some `<input>` and you are checking it with all these if statements where the final result is that you say that the data is "good" or "safe", then if one of your if statements doesn't work, then you might say the data is "safe" when it might not be.

If, instead of checking input, you just take input and combine it with other pieces of data you wrote to produce a final output that you believe is "safe" or safer, then I think it reduces the risk of missing a conditional.

For example, let's say there are a variety of images that we support. The user could provide a full URL, like `data:image/png;base64,<data>` and we could check to make sure the URL matches our conditions and if so, allow the use of that URL in the application. Or the user could provide a content-type, like `image/png` and `<data>` and we can look up in an object (`{'image/png': 'data:image/png;base64,'}`), the URL `prefix` for a given content-type and then construct the URL with that `prefix` and `<data>`, like `'<prefix><data>'`, or in this case, `'data:image/png;base64,<data>'`. Because we are only using the prefixes that we wrote (as long as we make sure the code actually uses the prefix), the prefix will only ever be a value we wrote or `undefined` if the content-type doesn't exist (in which case we can throw an error).

Reply:
It sounds like you've thought about this and have made all the changes you wanted, so let's resolve this!

Reply:
I added some additional tests around the existing validation